Lib/test/test_cgi.py

```diff
@@ -381,6 +381,51 @@ def testQSAndUrlEncode(self):
         v = gen_result(data, environ)
         self.assertEqual(self._qs_result, v)
 
+    def test_max_num_fields(self):
+        # For application/x-www-form-urlencoded
+        data = '&'.join(['a=a']*11)
+        environ = {
+            'CONTENT_LENGTH': str(len(data)),
+            'CONTENT_TYPE': 'application/x-www-form-urlencoded',
+            'REQUEST_METHOD': 'POST',
+        }
+
+        with self.assertRaises(ValueError):
+            form = cgi.FieldStorage(
+                fp=BytesIO(data.encode('ascii')),
+                environ=environ,
+                max_num_fields=10,
+            )
+
+        # For multipart/form-data
+        data = """---123
+Content-Disposition: form-data; name="a"
+
+a
+---123
+Content-Disposition: form-data; name="a"
+
+a
+---123
+Content-Disposition: form-data; name="a"
+
+a
+---123--
+"""
+        environ = {
+            'CONTENT_LENGTH':   str(len(data)),
+            'CONTENT_TYPE':     'multipart/form-data; boundary=-123',
+            'QUERY_STRING':     'a=a&a=a',
+            'REQUEST_METHOD':   'POST',
+        }
+
+        with self.assertRaises(ValueError):
+            form = cgi.FieldStorage(
+                fp=BytesIO(data.encode('ascii')),
+                environ=environ,
+                max_num_fields=4,
```

Comment:
Given that there are only 3 parts in `data`, it seems as though this shouldn't raise:  do we have an off-by-one somewhere?

Reply:
@tseaver You're right there are three parts in data, but there are two extra parts in `QUERY_STRING`. This was to test that both are considered.